test/model/typing_status_test.dart

```diff
@@ -0,0 +1,278 @@
+
+      check(model).typistsByLocation.deepEquals({
+        groupNarrow: {otherUser.userId: anyTimer, eg.selfUser.userId: anyTimer}});
+
+      model.handleTypingEvent(
+        mkTypingEvent(groupNarrow, TypingOp.stop, otherUser.userId));
+      check(model).typistsByLocation.deepEquals({
+        groupNarrow: {eg.selfUser.userId: anyTimer}});
+      checkNotifiedOnce();
+    }));
+
+    test('remove typists from different narrows', () => awaitFakeAsync((async) async {
+      prepareModel();
+
+      model.handleTypingEvent(
+        mkTypingEvent(dmNarrow, TypingOp.start, otherUser.userId));
+      checkNotifiedOnce();
+      model.handleTypingEvent(
+        mkTypingEvent(groupNarrow, TypingOp.start, thirdUser.userId));
+      checkNotifiedOnce();
+      model.handleTypingEvent(
+        mkTypingEvent(topicNarrow, TypingOp.start, eg.selfUser.userId));
+      checkNotifiedOnce();
+
+      async.elapse(const Duration(seconds: 5));
+      check(model).typistsByLocation.deepEquals({
+        dmNarrow: {otherUser.userId: anyTimer},
+        groupNarrow: {thirdUser.userId: anyTimer},
+        topicNarrow: {eg.selfUser.userId: anyTimer}});
+
+      model.handleTypingEvent(
+        mkTypingEvent(groupNarrow, TypingOp.stop, thirdUser.userId));
+      check(model).typistsByLocation.deepEquals({
+        dmNarrow: {otherUser.userId: anyTimer},
+        topicNarrow: {eg.selfUser.userId: anyTimer}});
+      checkNotifiedOnce();
+
+      model.handleTypingEvent(
+        mkTypingEvent(dmNarrow, TypingOp.stop, otherUser.userId));
+      check(model).typistsByLocation.deepEquals({
+        topicNarrow: {eg.selfUser.userId: anyTimer}});
+      checkNotifiedOnce();
+
+      model.handleTypingEvent(
+        mkTypingEvent(topicNarrow, TypingOp.stop, eg.selfUser.userId));
+      check(model).typistsByLocation.isEmpty();
+      checkNotifiedOnce();
+    }));
+  });
+
+  group('cancelling old timer', () {
+    test('when typing stopped early', () => awaitFakeAsync((async) async {
+      prepareModel();
+
+      model.handleTypingEvent(
+        mkTypingEvent(dmNarrow, TypingOp.start, otherUser.userId));
+      var oldTimer = check(model).typistsByLocation[dmNarrow][otherUser.userId]
+        ..isActive.isTrue();
+      checkNotifiedOnce();
+      model.handleTypingEvent(
+        mkTypingEvent(dmNarrow, TypingOp.stop, otherUser.userId));
+      oldTimer.isActive.isFalse();
+      checkNotifiedOnce();
+    }));
+
+    test('when typing repeatedly started', () => awaitFakeAsync((async) async {
+      prepareModel();
+
+      model.handleTypingEvent(
+        mkTypingEvent(groupNarrow, TypingOp.start, otherUser.userId));
+      var oldTimer = check(model).typistsByLocation[groupNarrow][otherUser.userId]
+        ..isActive.isTrue();
+      checkNotifiedOnce();
+      model.handleTypingEvent(
+        mkTypingEvent(groupNarrow, TypingOp.start, otherUser.userId));
+      oldTimer.isActive.isFalse();
+      checkNotNotified();
+    }));
+  });
```

Comment:
This checks that the old timer was canceled. It seems like it would be helpful to also check that a new timer was started.